package/system/procd/files/nand.sh

```diff
@@ -237,6 +248,111 @@ nand_upgrade_ubifs() {
+		return 1
+	}
+	local mtd_info=$(/usr/sbin/mtdinfo $kern_mtd_dev)
+	local mtd_type=`echo "$mtd_info" | sed -n 's/^Type: *\(.\+\)$/\1/p'`
+	local mtd_unit_size=`echo "$mtd_info" | sed -n 's/^Minimum input\/output unit size\: *\([0-9]\+\) *bytes$/\1/p'`
+	[ -z "$mtd_type" -o -z "$mtd_unit_size" ] && return 1
+	if [ "$mtd_type" == "nand" ]; then
+		mtd_type="NND"
+	elif [ "$mtd_type" == "nor" ]; then
+		mtd_type="NOR"
+	else
+		return 1
+	fi
+	local res="$mtd_type$mtd_unit_size"
+	while [ ${#res} -lt 8 ]; do
+		mtd_unit_size="0$mtd_unit_size"
+		res="$mtd_type$mtd_unit_size"
+	done
+	echo "$mtd_type$mtd_unit_size"
+	return 0
+}
+
+# Flash MIKROTIK yaffs2 kernel part of combined image to kernel partition
+# See https://github.com/adron-s/kernel2minor source code for details
+nand_upgrade_mikrotik_yaffs2_kernel() {
+	local CI_BLKSZ=65536
+	local kernelpart=$CI_KERNPART
+	local kern_mtd_index="$(find_mtd_index $CI_KERNPART)"
+	local kern_mtd_dev="/dev/mtd$kern_mtd_index"
+	local kern_ib_offset=$(($CI_BLKSZ / 8))
+	local kern_magic_id=$(dd if="$1" bs=8 skip=$kern_ib_offset count=1 2>/dev/null)
+	local kern_platform_name=$(dd if="$1" bs=8 skip=$(($kern_ib_offset+1)) count=1 2>/dev/null)
+	local kern_info_block_size=$((0x$(dd if="$1" bs=8 skip=$(($kern_ib_offset+2)) count=1 2>/dev/null)))
+	local kern_block_size=$((0x$(dd if="$1" bs=8 skip=$(($kern_ib_offset+4)) count=1 2>/dev/null)))
+	local kern_blocks=$((0x$(dd if="$1" bs=8 skip=$(($kern_ib_offset+5)) count=1 2>/dev/null)))
+	local kern_align_size=$((0x$(dd if="$1" bs=8 skip=$(($kern_ib_offset+10)) count=1 2>/dev/null)))
+	local this_platform_name=$(get_mikrotik_sysupgrade_platform_name)
+	[ $? -ne 0 ] && return 1 #if platform_name return error
+	[ "$kern_magic_id" != "MIKROTIK" ] && {
+		echo "Error! This image is not for MIKROTIK!"
+		return 1
+	}
+	[ "$kern_platform_name" != "$this_platform_name" ] && {
+		echo "Error! Image is for MIKROTIK but not for this platform!"
+		echo "This platform := '$this_platform_name'"
+		echo "Image platform := '$kern_platform_name'"
+		return 1
+	}
+	[ "$kern_align_size" -ne "$CI_BLKSZ" ] && {
+		echo "Error ! Kernel image align size '$kern_align_size' != '$CI_BLKSZ' !"
+		echo "Kernel image is corrupted?"
+		return 1
+	}
+	local kern_headers_size=$(($kern_align_size + $kern_info_block_size))
+	[ $(($kern_headers_size % $kern_block_size)) -ne 0 ] && {
+		echo "Error ! Kernel headers(sysupgrade + info_block) % block_size != 0"
+		echo "   ---> ($kern_align_size + $kern_info_block_size) % $kern_block_size != 0"
+		echo "Kernel image is corrupted?"
+		return 1
+	}
+	local kern_headers_size_in_blocks=$((kern_headers_size / $kern_block_size))
+	#write kernel
+	[ "$kern_blocks" != 0 -a -n "$kern_mtd_index" -a -e "$kern_mtd_dev" ] && {
+		mtd erase "$kern_mtd_dev" && \
+		    dd if="$1" bs=$kern_block_size skip=$kern_headers_size_in_blocks \
+		    count=$kern_blocks 2>/dev/null | \
+		    nandwrite -o $kern_mtd_dev - && return 0
```

Comment:
Stylistic remark: I had to read this several times to spot the `return 0` - maybe consider just one final `return $?` two lines below.